src/wp-includes/blocks.php

```diff
@@ -964,26 +979,28 @@ function set_ignored_hooked_blocks_metadata( &$parsed_anchor_block, $relative_po
 
 	/** This filter is documented in wp-includes/blocks.php */
 	$hooked_block_types = apply_filters( 'hooked_block_types', $hooked_block_types, $relative_position, $anchor_block_type, $context );
-	if ( empty( $hooked_block_types ) ) {
+	/** This filter is documented in wp-includes/blocks.php */
+	$hooked_blocks = apply_filters( 'hooked_blocks', $hooked_blocks, $relative_position, $parsed_anchor_block, $context );
+	$hooked_blocks = array_merge( $hooked_blocks, $hooked_block_types );
```

Comment:
My suggestion would be to remove the filter from here (and only keep it in `insert_hooked_blocks`). `set_ignored_hooked_blocks_metadata` is the mechanism we use to tell Block Hooks not to (re-)insert a hooked block if the containing template has already been edited by the user in the Site Editor, and thus, the hooked block has become part of the template (or was removed from it; either way, we want Block Hooks to respect that).

As mentioned elsewhere, if our use case is static templates, and if we don't have any concrete plans to make them editable by the user, we shouldn't include blocks that are added by the `hooked_blocks` filter in `ignoredHookedBlocks`. (We might then want to rename the filter to `injected_blocks`, to distinguish it better from hooked blocks.) Specifically, this will allow us to avoid using block signatures as an identifier.

(I think that the latter would introduce significant complexity; I'll elaborate in a follow-up comment.)

Reply:
I'll try to give an example for the complexities that block signatures would introduce:

Let's assume a plugin wants to insert the [Mini Cart block](https://github.com/woocommerce/woocommerce/tree/de90c5f46e7d79bac8a1c9a9dd1bf500b68c39b4/plugins/woocommerce-blocks/assets/js/blocks/mini-cart) after the Page List block, but it wants it to display the "bag" style rather than the "cart". This would look about like this (modulo escaping and whatnot):

```
<!-- wp:page-list {"metadata":{"ignoredHookedBlocks":["woocommerce/mini-cart-{"miniCartIcon":"bag"}"]}} /-->
<!-- wp:woocommerce/mini-cart {"miniCartIcon":"bag"} /-->
```

![image](https://github.com/WordPress/wordpress-develop/assets/96308/34ff2d36-9696-4452-9641-0fa401ccedcc)

Now let's assume the user opens the Site Editor to change the icon to `bag-alt`:

```
<!-- wp:page-list {"metadata":{"ignoredHookedBlocks":["woocommerce/mini-cart-{"miniCartIcon":"bag"}"]}} /-->
<!-- wp:woocommerce/mini-cart {"miniCartIcon":"bag-alt"} /-->
```

Block Hooks will continue to compare the block and its attribute to the signature in `ignoredHookedBlocks`, which is now out of sync, so it'll wrongly conclude that it needs to inject the Mini Cart block, resulting in two instances:

![image](https://github.com/WordPress/wordpress-develop/assets/96308/d2d1c138-70ed-4571-953c-8ff7db7fd155)

We could probably circumvent the above by updating the `ignoredHookedBlocks` metadata upon saving (which we don't do now; it's not needed for `ignoredHookedBlocks` that contain only block types). It's less straight-forward than it might seem (and might involve some level of heuristics, since we cannot know for sure that a the block was inserted by Block Hooks in the first place, or added later by the user). This extra complexity is avoidable if we only apply the filter within `insert_hooked_blocks`.

Reply:
I'm not opposed to keeping this in `insert_hooked_blocks` but there's a good chance that if we keep the product editor in blocks, it will be editable at some point.  I think it might be good to solve this issue of generic blocks for core blocks insertion as well.

Do you have any ideas for how might solve that?  An ID stands out to me as an easy way of doing this, but I know that blocks has been averse to adding IDs to blocks on the server.

Reply:
> Do you have any ideas for how might solve that?  An ID stands out to me as an easy way of doing this, but I know that blocks has been averse to adding IDs to blocks on the server.

Yeah, it's tricky. I've tried to think about other alternatives, but so far, I keep coming back to [named block variations](https://github.com/WordPress/gutenberg/issues/43743). I'll try to elaborate why.

If we use attributes to distinguish blocks from each other -- as done in this PR by means of a block signature -- we most likely wouldn't want the _entirety_ of attributes to enter into the picture. This is both for "information parsimony" and UX reasons: We'd likely want to avoid including every attribute of a given block in its signature (including `style` various block-supports related attributes), as that could yield a fairly massive string of serialized attributes.

For the user, it would also mean that if they change _any_ of those attributes, Block Hooks will no longer recognize it as a hooked block, and re-insert the hooked block itself (see the [shopping cart icon example](https://github.com/WordPress/wordpress-develop/pull/6228#discussion_r1534416349) above, but think even "smaller" changes, such as changing a font size or so).

To solve this problem, I think we'd need to select an attribute -- or a set of attributes -- that we actually _want_ to distinguish one block from another, if they are of the same type. E.g. for the [Social Icon block](https://wordpress.org/documentation/article/social-icons/#how-to-add-social-icons-and-links), we might want to distinguish block instances based on their `service` attribute (which can be `facebook`, `x`, `tumblr`, etc).

But this means that we need to introduce a way for blocks to declare which attributes carry that kind of meaning. And AFAICT, this is pretty exactly what named block variations are about.